### PR TITLE
test(ivy): Add unit tests for `insertTStyleBinding`

### PR DESCRIPTION
**Pull request by mhevery**
Pulled out `insertTStyleValue` into a stand alone function which is easier to test. Added unit tests which assert that linked lists are correctly updated. This is a prerequisite for integration work.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/core/src/render3/styling/STYLING.md [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+  - `color: {{tmplColor}};`
+  - `color: {{tmplColor2}};`
+- `my-green-dir` sets `color: green;` (implied)
+
+The algorithm needs to concatenate all of the individual styles into a single string in [correct order](#Concatenation-Order). The resulting string is:
+```
+color: red; color: green: color: {{tmplColor2}}; color: {{tmplColor}};
+```
+
+The resulting string is than written into the DOM using `element.style.cssText` (or `elemet.className`.) We rely on the browser to dedup the extra styles.
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/styling/STYLING.md]: Is this still the case? I thought that we need to write just `color: {{tmplColor}}` as people are asserting on the exact string in their tests?

File: packages/core/src/render3/styling/style_binding_list.ts [hunk 2]
```diff
@@ -0,0 +1,497 @@
...
+
+    // If we don't have template bindings than `tail` is 0.
+    const hasNoTemplateBinding = tmplTail === 0;
+    // This is important to know because that means that the `head` can't point to the first
+    // template bindings (there are none.) Instead the head points to the tail of the template.
+    const templateTail = hasNoTemplateBinding ?
+        tmplHead :
+        getTStylingRangePrev(tData[tmplHead + 1] as TStylingRange);
+    const templateHead = hasNoTemplateBinding ? 0 : tmplHead;
+    tData[index + 1] = toTStylingRange(templateTail, templateHead);
```
Comment (kara) [inline packages/core/src/render3/styling/style_binding_list.ts]:
```suggestion
    tData[index + 1] = toTStylingRange(previousNode, nextNode);
```

**Review by kara — CHANGES_REQUESTED** [review]
First pass through (will go through duplicates section next)

File: packages/core/src/render3/styling/STYLING.md (hunk 1 above)
Comment (mhevery) [reply packages/core/src/render3/styling/STYLING.md]: No, this is out of date, but after discussion with @matsko we decided to remove the whole file.

File: packages/core/src/render3/styling/style_binding_list.ts [hunk 3]
```diff
@@ -0,0 +1,497 @@
...
+ * concatenation list which is `#002`.
+ *
+ * To do so we keep a linked list of all of the bindings which pertain to this element.
+ * Notice that the bindings are inserted in the order of execution, but the `TView.data` allows
+ * us to traverse them in the order of priority.
+ *
+ * |Idx|`TView.data`|`LView`          | Notes
+ * |---|------------|-----------------|--------------
+ * |...|            |                 |
+ * |10 |`null`      |`{color: '#001'}`| `ɵɵstyleMap('color', {color: '#001'})`
```
[the remark this answers is not in the document]
Comment (mhevery) [reply packages/core/src/render3/styling/style_binding_list.ts]: 👍 The table is really for the future me.

**Review by mhevery — COMMENTED** [review]
- I have addressed all of your comments
- I have added more tests and documentation 7f0a10d7b7

File: packages/core/src/render3/styling/style_binding_list.ts [hunk 4]
```diff
@@ -0,0 +1,560 @@
+/**
+* @license
+* Copyright Google Inc. All Rights Reserved.
+*
+* Use of this source code is governed by an MIT-style license that can be
+* found in the LICENSE file at https://angular.io/license
+*/
+
+import {stylePropNeedsSanitization, ɵɵsanitizeStyle} from '../../sanitization/sanitization';
+import {assertEqual, assertNotEqual, throwError} from '../../util/assert';
```
Comment (mhevery) [inline packages/core/src/render3/styling/style_binding_list.ts]:
```suggestion
import {assertEqual, throwError} from '../../util/assert';
```

File: packages/core/test/render3/styling_next/class_differ_spec.ts [hunk 5]
```diff
@@ -98,6 +98,35 @@ describe('class differ', () => {
...
+      expect(classIndexOf('_ A _', 'A', 0)).toEqual(2);
+    });
+
+    it('should not match on partial matches', () => {
+      expect(classIndexOf('ABC AB', 'AB', 0)).toEqual(4);
+      expect(classIndexOf('AB ABC', 'AB', 1)).toEqual(-1);
+      expect(classIndexOf('ABC BC', 'BC', 0)).toEqual(4);
+      expect(classIndexOf('BC ABC', 'BB', 1)).toEqual(-1);
+    });
+  });
```
Comment (pkozlowski-opensource) [inline packages/core/test/render3/styling_next/class_differ_spec.ts]: Solid tests @mhevery - so good that you've taken time to consider all those cases. Our future self thank you!

File: packages/core/src/render3/styling/style_binding_list.ts [hunk 6]
```diff
@@ -0,0 +1,560 @@
...
+ * @param tNode `TNode` associated with the styling element.
+ * @param tStylingKey See `TStylingKey`.
+ * @param index location of where `tStyleValue` should be stored (and linked into list.)
+ * @param isHost `true` if the insertion is for a `hostBinding`. (insertion is in front of
+ *               template.)
+ * @param isClassBinding True if the associated `tStylingKey` as a `class` styling.
+ *                       `tNode.classBindings` should be used (or `tNode.styleBindings` otherwise.)
+ */
+export function insertTStylingBinding(
+    tData: TData, tNode: TNode, tStylingKey: TStylingKey, index: number, isHost: boolean,
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/styling/style_binding_list.ts]: nit: maybe rename `isHost` to `isHostBinding`? We use `isHost` quite often to denote a node that is a host of a directive (and does has to have host _bindings_) so things might get slightly confusing?

File: packages/core/src/render3/interfaces/node.ts [hunk 7]
```diff
@@ -599,6 +599,36 @@ export interface TNode {
    * will be placed into the initial styling slot in the newly created `TStylingContext`.
    */
   classes: StylingMapArray|TStylingContext|null;
+
+  /**
+   * Stores the head/tail of the class bindings.
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/interfaces/node.ts]: question: "Stores the head/tail of the class bindings." - we are storing _indexes_ of the head / tail, right? And those indexes point to entries in LView / TData ?

File: packages/core/src/render3/interfaces/node.ts [hunk 8]
```diff
@@ -599,6 +599,36 @@ export interface TNode {
...
+   *
+   * - If no bindings, the head and tail will both be 0.
+   * - If there are template bindings, stores the head/tail of the class bindings in the template.
+   * - If no template bindings but there are host bindings, the head value will point to the last
+   *   host binding for "class" (not the head of the linked list), tail will be 0.
+   *
+   * See: `style_binding_list.md` for details.
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/interfaces/node.ts]: docs nit: The `style_binding_list.md` file was removed, right? So we should remove this reference as well?

File: packages/core/src/render3/interfaces/node.ts [hunk 9]
```diff
@@ -599,6 +599,36 @@ export interface TNode {
...
+
+  /**
+   * Stores the head/tail of the style bindings.
+   *
+   * - If no bindings, the head and tail will both be 0.
+   * - If there are template bindings, stores the head/tail of the style bindings in the template.
+   * - If no template bindings but there are host bindings, the head value will point to the last
+   *   host binding for "style" (not the head of the linked list), tail will be 0.
+   *
+   * See: `style_binding_list.md` for details.
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/interfaces/node.ts]: docs nit: The `style_binding_list.md` file was removed, right? So we should remove this reference as well?

File: packages/core/test/render3/styling_next/style_binding_list_spec.ts [hunk 10]
```diff
@@ -0,0 +1,659 @@
...
+        null, null,                      // 26
+        null, null,                      // 28
+        null, null,                      // 30
+      ]);
+      expectPriorityOrder(tData, tNode, false).toEqual([
+        [10, null, false, true],     // 10 - Template:  ɵɵstyleMap({color: '#001'});
+        [12, 'color', true, false],  // 12 - Template:  ɵɵstyleProp('color', '#002'});
+      ]);
+
+      debugger;
```
Comment (pkozlowski-opensource) [inline packages/core/test/render3/styling_next/style_binding_list_spec.ts]: lint: `debugger;`

File: packages/core/src/render3/interfaces/node.ts (hunk 8 above)
Comment (kara) [reply packages/core/src/render3/interfaces/node.ts]: I think he means `style_binding_list.ts` here. But we should update the name!

File: packages/core/src/render3/interfaces/styling.ts [hunk 11]
```diff
@@ -471,3 +472,173 @@ export const enum StylingMapsSyncMode {
...
+ *
+ * We use `TStylingRange` to store the previous and next index into the `TData` where the template
+ * bindings can be found.
+ *
+ * - bit 0 is used to mark that the previous index has a duplicate for current value.
+ * - bit 1 is used to mark that the next index has a duplicate for the current value.
+ * - bits 2-16 are used to encode the next/tail of the template.
+ * - bits 17-32 are used to encode the previous/head of template.
+ *
+ * NODE: *duplicate* false implies that it is statically know that this binding will not collide
```
Comment (kara) [inline packages/core/src/render3/interfaces/styling.ts]:
```suggestion
 * NODE: *duplicate* false implies that it is statically known that this binding will not collide
```

File: packages/core/src/render3/interfaces/styling.ts [hunk 12]
```diff
@@ -471,3 +472,173 @@ export const enum StylingMapsSyncMode {
...
+ * with other binding and therefor there is no need to check other bindings. For example the
```
Comment (kara) [inline packages/core/src/render3/interfaces/styling.ts]:
```suggestion
 * with other bindings and therefore there is no need to check other bindings. For example the
```

File: packages/core/src/render3/styling/style_binding_list.ts [hunk 13]
```diff
@@ -0,0 +1,560 @@
...
+
+/**
+ * Insert new `tStyleValue` at `TData` and link existing style bindings such that we maintain linked
+ * list of styles and compute the duplicate flag.
+ *
+ * Note: this function is executed during `firstUpdatePass` only to populate the `TView.data`.
+ *
+ * The function works by keeping track of `tStylingRange` which contains two pointers pointing to
+ * the head/tail of the template portion of the styles.
+ *  - if `isHost === false` (we are template) than insertion is at tail of `TStylingRange`
```
Comment (kara) [inline packages/core/src/render3/styling/style_binding_list.ts]:
```suggestion
 *  - if `isHost === false` (we are template) then insertion is at tail of `TStylingRange`
```

File: packages/core/src/render3/styling/style_binding_list.ts [hunk 14]
```diff
@@ -0,0 +1,560 @@
...
+ *  - if `isHost === true` (we are host binding) than insertion is at head of `TStylingRange`
```
Comment (kara) [inline packages/core/src/render3/styling/style_binding_list.ts]:
```suggestion
 *  - if `isHost === true` (we are host binding) then insertion is at head of `TStylingRange`
```

File: packages/core/src/render3/styling/style_binding_list.ts [hunk 15]
```diff
@@ -0,0 +1,560 @@
...
+                   'Should be called during \'firstUpdatePass` only.');
+  let tBindings = isClassBinding ? tNode.classBindings : tNode.styleBindings;
+  let tmplHead = getTStylingRangePrev(tBindings);
+  let tmplTail = getTStylingRangeNext(tBindings);
+
+  tData[index] = tStylingKey;
+  if (isHost) {
+    // We are inserting host bindings
+
+    // If we don't have template bindings than `tail` is 0.
```
Comment (kara) [inline packages/core/src/render3/styling/style_binding_list.ts]:
```suggestion
    // If we don't have template bindings then `tail` is 0.
```

File: packages/core/src/render3/styling/style_binding_list.ts [hunk 16]
```diff
@@ -0,0 +1,560 @@
...
+    }
+  }
+  if (foundDuplicate) {
+    tData[index + 1] = isPrevDir ? setTStylingRangePrevDuplicate(tStylingAtIndex) :
+                                   setTStylingRangeNextDuplicate(tStylingAtIndex);
+  }
+}
+
+/**
+ * Computes the new styling value starting `index` styling binding.
```
Comment (kara) [inline packages/core/src/render3/styling/style_binding_list.ts]:
```suggestion
 * Computes the new styling value starting at `index` styling binding.
```

File: packages/core/src/render3/styling/style_binding_list.ts [hunk 17]
```diff
@@ -0,0 +1,560 @@
...
+ * @param lView `LView` containing the styling values.
+ *              - `LView[index]` contains the binding value.
+ *              - `LView[index + 1]` contains the concatenated value up to this point.
+ * @param index the location in `TData`/`LView` where the styling search should start.
+ * @param isClassBinding `true` if binding to `className`; `false` when binding to `style`.
+ */
+export function flushStyleBinding(
+    tData: TData, tNode: TNode, lView: LView, index: number, isClassBinding: boolean): string {
+  const tStylingRangeAtIndex = tData[index + 1] as TStylingRange;
+  const lastGoodValueIndex = getTStylingRangePrev(tStylingRangeAtIndex);
```
Comment (kara) [inline packages/core/src/render3/styling/style_binding_list.ts]: What's a "good value"? I'm guessing you mean the last concatenated string value composed of styles that haven't changed? Can we add a comment or update the name?

File: packages/core/src/render3/styling/style_binding_list.ts [hunk 18]
```diff
@@ -0,0 +1,560 @@
...
+/**
+ * Append new styling to the currently concatenated styling text.
+ *
+ * This function concatenates the existing `className`/`cssText` text with the binding value.
+ *
+ * @param text Text to concatenate to.
+ * @param stylingKey `TStylingKey` holding the key (className or style property name).
+ * @param value The value for the key.
+ *         - `isClassBinding === true`
+ *              - `boolean` if `true` than add the key to the class list string.
```
Comment (kara) [inline packages/core/src/render3/styling/style_binding_list.ts]:
```suggestion
 *              - `boolean` if `true` then add the key to the class list string.
```

File: packages/core/src/render3/styling/style_binding_list.ts [hunk 19]
```diff
@@ -0,0 +1,560 @@
...
+ *              - `Object` add object key/value to the styles.
+ * @param sanitizer Optional sanitizer to use. If `null` the `stylingKey` sanitizer will be used.
+ *        This is provided so that `ɵɵstyleMap()`/`ɵɵclassMap()` can recursively call
+ *        `appendStyling` without having ta package the sanitizer into `TStylingSanitizationKey`.
+ * @param hasPreviousDuplicate determines if there is a chance of duplicate.
+ *         - `true` the existing `text` should be searched for duplicates and if any found they
+ *           should be removed.
+ *         - `false` Fast path, just concatenate the strings.
+ * @param isClassBinding Determines if the `text` is `className` or `cssText`.
+ */
```
Comment (kara) [inline packages/core/src/render3/styling/style_binding_list.ts]: Add a return value line to the comment?

File: packages/core/src/render3/styling/style_binding_list.ts [hunk 20]
```diff
@@ -0,0 +1,560 @@
...
+    }
+    return text;
+  }
+};
+
+/**
+ * `ɵɵstyleMap()` inserts `STYLE_MAP_STYLING_KEY` as a key to the `insertTStylingBinding()`.
+ *
+ * The purpose of this key is to add style map abilities to the concatenation in a tree shakable
+ * way. If `ɵɵstyleMap()` is not referenced than `STYLE_MAP_STYLING_KEY` will become eligible for
```
Comment (kara) [inline packages/core/src/render3/styling/style_binding_list.ts]:
```suggestion
 * way. If `ɵɵstyleMap()` is not referenced then `STYLE_MAP_STYLING_KEY` will become eligible for
```

File: packages/core/src/render3/styling/style_binding_list.ts [hunk 21]
```diff
@@ -0,0 +1,560 @@
...
+ * tree shaking. (`STYLE_MAP_STYLING_KEY` also pulls int the sanitizer as `ɵɵstyleMap()` could have
```
Comment (kara) [inline packages/core/src/render3/styling/style_binding_list.ts]:
```suggestion
 * tree shaking. (`STYLE_MAP_STYLING_KEY` also pulls in the sanitizer as `ɵɵstyleMap()` could have
```

File: packages/core/src/render3/styling/style_binding_list.ts [hunk 22]
```diff
@@ -0,0 +1,560 @@
...
+            (value, key) => text = appendStyling(
+                text, key, value.old, stylePropNeedsSanitization(key) ? ɵɵsanitizeStyle : null,
+                true, false));
+      } else {
+        // No duplicates, just append it.
+        text = text === '' ? value : text + '; ' + value;
+      }
+    } else {
+      // All other cases are not supported.
+      ngDevMode && throwError('Unsupported value for class binding: ' + value);
```
Comment (kara) [inline packages/core/src/render3/styling/style_binding_list.ts]:
```suggestion
      ngDevMode && throwError('Unsupported value for style binding: ' + value);
```

File: packages/core/src/render3/styling/style_binding_list.ts [hunk 23]
```diff
@@ -0,0 +1,560 @@
...
+ * This key supports: `strings`, and `object` (as Map). In each case it is necessary to
+ * break the style into parts and concatenate the parts into the `text`. The concatenation needs
+ * to be done in parts as each key is individually subject to overwrites.
+ */
+export const STYLE_MAP_STYLING_KEY: TStylingMapKey = {
+  key: null,
+  extra: (text: string, value: any, hasPreviousDuplicate: boolean): string => {
+    if (Array.isArray(value)) {
+      // We don't support Arrays
+      ngDevMode && throwError('Style binding do not support arrays bindings: ' + value);
```
Comment (kara) [inline packages/core/src/render3/styling/style_binding_list.ts]:
```suggestion
      ngDevMode && throwError('Style bindings do not support array bindings: ' + value);
```

File: packages/core/src/render3/interfaces/node.ts (hunk 7 above)
Comment (mhevery) [reply packages/core/src/render3/interfaces/node.ts]: yes, rewrote to make it more clear.

File: packages/core/src/render3/interfaces/node.ts (hunk 8 above)
Comment (mhevery) [reply packages/core/src/render3/interfaces/node.ts]: fixed

File: packages/core/src/render3/interfaces/node.ts (hunk 9 above)
Comment (mhevery) [reply packages/core/src/render3/interfaces/node.ts]: fixed

File: packages/core/src/render3/styling/style_binding_list.ts (hunk 6 above)
Comment (mhevery) [reply packages/core/src/render3/styling/style_binding_list.ts]: renamed

File: packages/core/src/render3/styling/style_binding_list.ts (hunk 17 above)
Comment (mhevery) [reply packages/core/src/render3/styling/style_binding_list.ts]: renamed and added comment.

File: packages/core/src/render3/styling/style_binding_list.ts (hunk 19 above)
Comment (mhevery) [reply packages/core/src/render3/styling/style_binding_list.ts]: added.

File: packages/core/test/render3/styling_next/class_differ_spec.ts (hunk 5 above)
Comment (mhevery) [reply packages/core/test/render3/styling_next/class_differ_spec.ts]: Thank you.

File: packages/core/test/render3/styling_next/style_binding_list_spec.ts (hunk 10 above)
Comment (mhevery) [reply packages/core/test/render3/styling_next/style_binding_list_spec.ts]: removed

File: packages/core/src/render3/interfaces/styling.ts [hunk 24]
```diff
@@ -471,3 +472,173 @@ export const enum StylingMapsSyncMode {
...
+  /**
+   * This bit is set if the previous bindings contains a binding which could possibly cause a
+   * duplicate. For example: `<div [style]="map" [style.width]="width">`, the `width` binding will
+   * have previous duplicate set. The implication is that if `width` binding becomes `null`, it is
+   * necessary to defer the value to `map.width`. (Because `width` overwrites `map.width`.)
+   */
+  PREV_DUPLICATE = 0x02,
+
+  /**
+   * This bit is set to if the next binding contains a binding which could possibly cause a
```
Comment (kara) [inline packages/core/src/render3/interfaces/styling.ts]:
```suggestion
   * This bit is set if the next binding contains a binding which could possibly cause a
```

File: packages/core/src/render3/styling/style_binding_list.ts (hunk 20 above)
Comment (kara) [reply packages/core/src/render3/styling/style_binding_list.ts]: Update? ^

File: packages/core/src/render3/styling/style_binding_list.ts [hunk 25]
```diff
@@ -0,0 +1,563 @@
...
+    text = text === '' ? keyValue : text + '; ' + keyValue;
+  }
+  return text;
+}
+
+/**
+ * `ɵɵclassMap()` inserts `CLASS_MAP_STYLING_KEY` as a key to the `insertTStylingBinding()`.
+ *
+ * The purpose of this key is to add class map abilities to the concatenation in a tree shakable
+ * way. If `ɵɵclassMap()` is not referenced than `CLASS_MAP_STYLING_KEY` will become eligible for
```
Comment (kara) [inline packages/core/src/render3/styling/style_binding_list.ts]:
```suggestion
 * way. If `ɵɵclassMap()` is not referenced then `CLASS_MAP_STYLING_KEY` will become eligible for
```

File: packages/core/src/render3/styling/style_binding_list.ts [hunk 26]
```diff
@@ -0,0 +1,563 @@
...
+ * @param lView `LView` containing the styling values.
+ *              - `LView[index]` contains the binding value.
+ *              - `LView[index + 1]` contains the concatenated value up to this point.
+ * @param index the location in `TData`/`LView` where the styling search should start.
+ * @param isClassBinding `true` if binding to `className`; `false` when binding to `style`.
+ */
+export function flushStyleBinding(
+    tData: TData, tNode: TNode, lView: LView, index: number, isClassBinding: boolean): string {
+  const tStylingRangeAtIndex = tData[index + 1] as TStylingRange;
+  // When styling changes we don't have to start at the begging. Instead we start at the change
```
Comment (kara) [inline packages/core/src/render3/styling/style_binding_list.ts]:
```suggestion
  // When styling changes we don't have to start at the beginning. Instead we start at the first changed
```

**Review by kara — APPROVED** [review]
LGTM, but a few more typos to fix

**Comment by mhevery** [comment]
merged as a71965671c

**Pull request closed without merging**